From 36e7e9cac9a6cea03c5272aaf628b874840a5fb2 Mon Sep 17 00:00:00 2001
From: tommy-huang <tommy_huang@aspeedtech.com>
Date: Wed, 1 Dec 2021 16:24:28 +0800
Subject: [PATCH 02/10] drivers:i2c:fix the stop condition

Add virtual slave stop condition.
Add more DMA fire condition.

Signed-off-by: tommy-huang <tommy_huang@aspeedtech.com>
Change-Id: I095836b49ea5c6be446bc1814fbc7216dc4172bf
---
 drivers/i2c/i2c_aspeed.c | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/i2c_aspeed.c b/drivers/i2c/i2c_aspeed.c
index b8222da0b6..5daff7199f 100644
--- a/drivers/i2c/i2c_aspeed.c
+++ b/drivers/i2c/i2c_aspeed.c
@@ -1256,6 +1256,32 @@ int aspeed_i2c_slave_irq(const struct device *dev)
 		sts &= ~(AST_I2CS_PKT_DONE | AST_I2CS_PKT_ERROR);
 		sys_write32(AST_I2CS_PKT_DONE, i2c_base + AST_I2CS_ISR);
 		switch (sts) {
+		case AST_I2CS_SLAVE_MATCH:
+		case AST_I2CS_SLAVE_MATCH | AST_I2CS_RX_DONE | AST_I2CS_Wait_RX_DMA:
+		case AST_I2CS_SLAVE_MATCH | AST_I2CS_Wait_RX_DMA:
+			LOG_DBG("S : Sw\n");
+			if (slave_cb->write_requested) {
+				slave_cb->write_requested(data->slave_cfg);
+			}
+			if (sys_read32(i2c_base + AST_I2CM_ISR) & AST_I2CS_RX_DMA_EN)
+				LOG_DBG("S : Sw|D - Wait normal\n");
+			else {
+				LOG_DBG("S : Sw|D - Issue rx dma\n");
+				cmd = AST_I2CS_ACTIVE_ALL | AST_I2CS_PKT_MODE_EN;
+				if (config->mode == DMA_MODE) {
+					cmd |= AST_I2CS_RX_DMA_EN;
+					sys_write32(AST_I2CS_SET_RX_DMA_LEN(I2C_SLAVE_BUF_SIZE)
+					, i2c_base + AST_I2CS_DMA_LEN);
+				} else if (config->mode == BUFF_MODE) {
+					cmd |= AST_I2CS_RX_BUFF_EN;
+					sys_write32(AST_I2CC_SET_RX_BUF_LEN(I2C_BUF_SIZE)
+					, i2c_base + AST_I2CC_BUFF_CTRL);
+				} else {
+					cmd &= ~AST_I2CS_PKT_MODE_EN;
+				}
+				sys_write32(cmd, i2c_base + AST_I2CS_CMD_STS);
+			}
+			break;
 		case AST_I2CS_SLAVE_MATCH | AST_I2CS_STOP:
 			LOG_DBG("S : Sw | P\n");
 			if (slave_cb->stop) {
@@ -1280,13 +1306,15 @@ int aspeed_i2c_slave_irq(const struct device *dev)
 		case AST_I2CS_RX_DONE | AST_I2CS_Wait_RX_DMA | AST_I2CS_STOP:
 		/*LOG_DBG(data->dev, "S : D|P wait rx dma workaround\n");*/
 		case AST_I2CS_RX_DONE | AST_I2CS_STOP:
-		case AST_I2CS_SLAVE_MATCH | AST_I2CS_RX_DONE | AST_I2CS_Wait_RX_DMA:
 		case AST_I2CS_SLAVE_MATCH | AST_I2CS_RX_DONE | AST_I2CS_STOP:
 			if (sts & AST_I2CS_STOP) {
 				if (sts & AST_I2CS_SLAVE_MATCH) {
 					LOG_DBG("S : Sw|D|P\n");
 				} else {
 					LOG_DBG("S : D|P\n");
+					if (slave_cb->stop) {
+						slave_cb->stop(data->slave_cfg);
+					}
 				}
 			} else {
 				LOG_DBG("S : Sw|D\n");
-- 
2.17.1

